actors/miner/src/lib.rs

```diff
@@ -1031,136 +1031,30 @@ impl Actor {
-                    update.sector_number,
-                );
-                continue;
-            }
-
-            if update.deals.is_empty() {
-                info!("must have deals to update, skipping sector {}", update.sector_number,);
-                continue;
-            }
-
-            if update.deals.len() as u64 > sector_deals_max(rt.policy(), info.sector_size) {
-                info!("more deals than policy allows, skipping sector {}", update.sector_number,);
-                continue;
-            }
-
-            if update.deadline >= rt.policy().wpost_period_deadlines {
-                info!(
-                    "deadline {} not in range 0..{}, skipping sector {}",
-                    update.deadline,
-                    rt.policy().wpost_period_deadlines,
-                    update.sector_number
-                );
-                continue;
-            }
-
-            // Skip checking if CID is defined because it cannot be so in Rust
-
-            if !is_sealed_sector(&update.new_sealed_cid) {
-                info!(
-                    "new sealed CID had wrong prefix {}, skipping sector {}",
-                    update.new_sealed_cid, update.sector_number
-                );
-                continue;
-            }
+        let mut sectors = Sectors::load(&store, &state.sectors)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load sectors array")?;
 
-            // If the deadline is the current or next deadline to prove, don't allow updating sectors.
-            // We assume that deadlines are immutable when being proven.
-            if !deadline_is_mutable(
-                rt.policy(),
-                state.current_proving_period_start(rt.policy(), rt.curr_epoch()),
-                update.deadline,
-                rt.curr_epoch(),
-            ) {
-                info!(
-                    "cannot upgrade sectors in immutable deadline {}, skipping sector {}",
-                    update.deadline, update.sector_number
-                );
-                continue;
-            }
-
-            if !state
-                .check_sector_active(
-                    rt.policy(),
-                    rt.store(),
-                    update.deadline,
-                    update.partition,
-                    update.sector_number,
-                    true,
-                )
-                .map_err(|_| actor_error!(illegal_argument, "error checking sector health"))?
-            {
-                info!("sector isn't healthy, skipping sector {}", update.sector_number);
-                continue;
-            }
-
-            let res = Sectors::must_get(&sectors, update.sector_number);
-            let sector_info = if let Ok(value) = res {
-                value
-            } else {
-                info!("failed to get sector, skipping sector {}", update.sector_number);
-                continue;
-            };
-
-            if !sector_info.deal_ids.is_empty() {
```

Comment:
This should not block this PR but I think this points out that we've missed something critical for sector content added in relation to snap deals.

Without deal ids attached to the sector info in the new flow we will not be able to distinguish between sectors with data and sectors without data.

Without deal ids we need a boolean flag on sector info to indicate that it bears data.  We can take the opportunity to make [an extendable flags field](https://github.com/filecoin-project/FIPs/discussions/601).

We could also achieve this by enforcing storing of CommD for all non zero sectors on sector info.  IMO this would be a good thing but the state cost is a probably high and this is more contentious.

Reply:
Note this shouldn't stop us from total removal of deal_ids during migration -- we can just set the flag then depending on existance of deal ids.

If we went the required commd route we would probably want to keep deal ids for old non-CC sectors but we could also remove them and use the flag for fallback.

Reply:
Thanks for noticing this. I was dimly aware I was punting something when I moved this code. We might have easier paths to distinguishing them than requiring a new flag:
- deal weight is currently set on all data in sectors, and we can preserve that; we can reject updates to non-zero deal weight
- sector key CID is only set on the first replica update; we can reject updates to non-None sector key

Reply:
Nice I had forgotten deal weight